Added removeValue method to ArrayHelper

nilsburg opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes

New method in ArrayHelper to remove an item by value.

```php
// $array = ['value1', 'value2'];
$key = \yii\helpers\ArrayHelper::removeValue($array, 'value2');
// result:
// $array = ['value1']

dynasource commented:
this is a usefull method for ArrayHelper. Can you add the unit tests to your PR to show it works correctly?

nilsburg commented:
Added the unit test. I don't have much experience with unit tests so let me know if I need to change something.
Thanks!

dynasource commented:
Looks good for a rookie ;)  Can you update the changelog and reset your commits to 1 commit? Thanks

nilsburg commented:
Hi,
I've updated the CHANGELOG, but I've no idea how to reset the commits to 1. Would you mind guiding me? thanks!

dynasource commented:
this is a way:
* soft reset your branch to the last commit from the repo (before your commits)
* commit again
* force push

nilsburg commented:
Done!
Thank you for the help!

dynasource commented on framework/CHANGELOG.md:
```diff
@@ -17,7 +17,7 @@ Yii Framework 2 Change Log
 - Bug #9796: Initialization of not existing `yii\grid\ActionColumn` default buttons (arogachev)
 - Enh: Added constants for specifying `yii\validators\CompareValidator::$type` (cebe)
-
+- Enh: Added `removeValue` method to `yii\yii\helpers\BaseArrayHelper` (nilsburg)
```
incorrect namespace yii\yii\helpers\BaseArrayHelper

dynasource commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,32 @@ public static function remove(&$array, $key, $default = null)
     }
 
     /**
+     * Search and removes an item from an array and returns the key if found. If the value does not exists it will return the default value.
```
please use multiple lines

dynasource commented on framework/helpers/BaseArrayHelper.php:
````diff
@@ -247,6 +247,32 @@ public static function remove(&$array, $key, $default = null)
+     * Example,
+     *
+     * ```php
+     * // $array = ['value1', 'value2'];
````
remove the //

dynasource commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,32 @@ public static function remove(&$array, $key, $default = null)
+     * // $array = ['value1', 'value2'];
+     * $key = \yii\helpers\ArrayHelper::removeValue($array, 'value2');
+     * // result:
+     * // $array = ['value1']
```
remove the //

nilsburg replied:
I think that I should leave the // on this line since it represents the result. Please let me know if I am wrong.

dynasource replied:
correct. I guess these // dont really add to the readability ;), but it is consistent with the other methods.

dynasource commented:
sorry, still saw some improvements

nilsburg commented:
Ok, I've just pushed the changes. Sorry for the mistakes!

dynasource commented:
thanks! Now lets wait for others @yiisoft/core-developers opinions

rob006 commented:
What is the use case for `$default`?

samdark commented on framework/CHANGELOG.md:
```diff
@@ -17,7 +17,7 @@ Yii Framework 2 Change Log
 - Bug #9796: Initialization of not existing `yii\grid\ActionColumn` default buttons (arogachev)
 - Enh: Added constants for specifying `yii\validators\CompareValidator::$type` (cebe)
-
+- Enh: Added `removeValue` method to `yii\helpers\BaseArrayHelper` (nilsburg)
```
Need issue number here i.e. `Enh #12881`.

samdark commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,33 @@ public static function remove(&$array, $key, $default = null)
+     * @param null $default
+     * @return mixed|null
+     */
+    public static function removeValue(&$array, $value, $default = null)
```
I think naming could be improved because `removeValue` sounds like `unset` and nothing more. How about `pullValue`?

nilsburg replied:
I think `removeValue` tells you exactly what the function does, I agree that `pullValue` could work, but I think it is confusing because `pull` can be confused by just retrieving some value from the array, like `git pull` which doesn't remove anything.

rob006 replied:
It should be named `popValue`.

nilsburg replied:
Could it not be confused by the functionality of the `array_pop` function and expect that it will only remove the last occurrence? I've no problem renaming it, but in my opinion removeValue tells you exactly what it does.

rob006 replied:
This method is useful only because returns keys. IMO removing elements from array is not the main functionality of this method - if I want only remove something, I just use native `array_diff()`.

nilsburg replied:
the main reason I came up with this method was because I needed to remove only one value from an array, I think this method does the job and is easier to use than `array_diff()`.

samdark commented on framework/helpers/BaseArrayHelper.php:
````diff
@@ -247,6 +247,33 @@ public static function remove(&$array, $key, $default = null)
+     * // $array = ['value1']
+     * ```
+     *
+     * @param $array
````
Need description.

samdark commented on framework/helpers/BaseArrayHelper.php:
````diff
@@ -247,6 +247,33 @@ public static function remove(&$array, $key, $default = null)
+     * ```
+     *
+     * @param $array
+     * @param $value
````
Need description.

samdark commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,33 @@ public static function remove(&$array, $key, $default = null)
+     *
+     * @param $array
+     * @param $value
+     * @param null $default
```
Need description.

samdark commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,33 @@ public static function remove(&$array, $key, $default = null)
+     * @param $value
+     * @param null $default
+     * @return mixed|null
+     */
```
Need `@since 2.0.11` tag here.

samdark commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,33 @@ public static function remove(&$array, $key, $default = null)
+     */
+    public static function removeValue(&$array, $value, $default = null)
+    {
+        if(is_array($array) && ($key = array_search($value, $array)) !== false){
```
Formatting is wrong. Should be `if (is_array`

samdark commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,33 @@ public static function remove(&$array, $key, $default = null)
+    {
+        if(is_array($array) && ($key = array_search($value, $array)) !== false){
+            unset($array[$key]);
+            return $key;
```
Shouldn't it return value instead of key?

Faryshta replied:
whats the use of returning the value you sent to begin with?

nilsburg replied:
I think that returning the key can be useful in case you want to store the removed key somewhere.

samdark reviewed (changes requested):
`composer.lock` should not be committed.

Overall it looks OK except the fact that it returns key which is kinda useless and not value which is useful.

dynasource commented:
thinking more about this method, there is a problem. There can be more similar values in an array. Currently only the first one is removed. Of course all occurrences should be unset.

Regarding the return value. There are usecases thinkable. Lets say you have an array with firstnames => surnames. You want to remove a family from the array, but get all the first names. This implies we would expect the method to return all the first names.

nilsburg commented:
Oh, sorry, I think I closed the PR by mistake... is there a way to reopen it?

dynasource commented:
dont you have a button reopen?

nilsburg commented:
Yes, but it is greyed out because there are no new commits. Do you think if I reset and commit again it would be active again?
Damm I I feel so dumb!

dynasource commented:
No problem. I cant either :) so I guess we need new commits.

Regarding the new commits, please take a look at https://github.com/yiisoft/yii2/pull/12881#issuecomment-257509054

nilsburg commented:
About similar values in the array. I think it is the user who would have to implement that in his code to remove all the similar values, but it is possible to add an option to remove all the occurrences... I am not sure which is the best option.

dynasource commented:
removing only the first one is rather arbitrary

nilsburg commented:
The thing is that if I remove all the occurrences and I want to store the removed keys, should I return an array with the removed keys?

dynasource commented:
correct. Your method should be altered and return an array at all times. This would be consistent, also with respect to future return type hinting of PHP7

arogachev commented:
@dynasource Now Git has an option of squashing commits with merge so manual reset to 1 commit is not required. Check this [Github blog bost](https://github.com/blog/2141-squash-your-commits). I think it's already configured properly because I saw that @samdark merded some PRs that way.

dynasource commented:
@arogachev, nice to know. Thanks for sharing!

samdark commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,40 @@ public static function remove(&$array, $key, $default = null)
+     * @param array $array the array where to look the value from
+     * @param string $value the value to remove from the array
+     * @param mixed $default the value to return in case the value to remove was not found
+     * @return mixed|null the key of the element if found, default value otherwise
```
But why returning keys? What's the use case for that?

dynasource replied:
plz see https://github.com/yiisoft/yii2/pull/12881#issuecomment-257509054

samdark replied:
That sounds very specific to me...

rob006 replied:
We could return values indexed by keys.

SilverFire reviewed (changes requested):
Code style should be adjusted

SilverFire commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,40 @@ public static function remove(&$array, $key, $default = null)
+    public static function removeValue(&$array, $value, $default = null)
+    {
+        $result = [];
+        if (is_array($array)) {
```
Will not work for ArrayAccess objects

SilverFire replied:
Does it make any sense to continue execution when the `$array` is not array indeed?

SilverFire commented on tests/framework/helpers/ArrayHelperTest.php:
```diff
@@ -142,6 +142,18 @@ public function testRemove()
+        $this->assertEquals($array, [0=>'value1', 2=>'value3']);
+
+        $default = ArrayHelper::removeValue($array, 'nonExisting', 'defaultValue');
+        $this->assertEquals('defaultValue', $default);
```
Wrong cases should be tested as well

nilsburg commented:
Hi,
Following the suggestions I've made some changes in the method.
I removed the $default parameter, I couldn't find really a good use for it.
It now returns the removed items instead of just the keys.
I've also modified the example following @dynasource example here: https://github.com/yiisoft/yii2/pull/12881#issuecomment-257509054
About the code styling, I applied the PSR1/PSR2 code styling with PhpStorm, hope it's ok now.

So let me know if you can think of more improvements and thanks for the suggestions.

samdark commented:
I think now it's ready. @dynasource, @SilverFire ?

samdark commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -247,6 +247,38 @@ public static function remove(&$array, $key, $default = null)
+     * $removed = \yii\helpers\ArrayHelper::removeValue($array, 'Jackson');
+     * // result:
+     * // $array = ['Bob'=>'Dylan', 'Mick'=>'Jagger'];
+     * // $removed = ['Michael'=>'Jackson', 'Janet'=>'Jackson'];
```
Yes. Indeed.

nilsburg replied:
Done

samdark commented on tests/framework/helpers/ArrayHelperTest.php:
```diff
@@ -142,6 +142,31 @@ public function testRemove()
         $this->assertEquals('defaultValue', $default);
     }
 
+    public function testRemoveValue()
```
The test method has two cases. Could you split it into test method and data-provider?

dynasource replied:
```php
    public function testRemoveValueMultiple()
    {
        $array = [
            'Bob' => 'Dylan',
            'Michael' => 'Jackson',
            'Mick' => 'Jagger',
            'Janet' => 'Jackson'
        ];

        $removed = ArrayHelper::removeValue($array, 'Jackson');

        $this->assertEquals([
            'Bob' => 'Dylan',
            'Mick' => 'Jagger'
        ], $array);
        $this->assertEquals([
            'Michael' => 'Jackson',
            'Janet' => 'Jackson'
        ], $removed);
    }

    public function testRemoveValueNotExisting()
    {
        $array = [
            'Bob' => 'Dylan',
            'Michael' => 'Jackson',
            'Mick' => 'Jagger',
            'Janet' => 'Jackson'
        ];

        $removed = ArrayHelper::removeValue($array, 'Marley');

        $this->assertEquals([
            'Bob' => 'Dylan',
            'Michael' => 'Jackson',
            'Mick' => 'Jagger',
            'Janet' => 'Jackson'
        ], $array);
        $this->assertEquals([], $removed);
    }
```

dynasource replied:
you can use the code above. Please note the following improvements:
* tests should be clear in 1 second. If you still have to think when you read it, its not ready yet
* 1 usecase per method. This enables you to create a good name for the method  
* try to follow the following 3 step template and split them with a blank line. It will be much easier to distinct.

1. prepare the data
2. execute the code
3. do the assertions

* dont put arrays in 1 line. Split them over multiple lines (otherwise you still dont see what has changed)
* dont reverse 'actual' and 'expected' in assertions. First you have the 'expected' and then 'actual':

Example:
```php public static function assertEquals($expected, $actual) ```

nilsburg replied:
@dynasource I've added your code for the tests. Thank you very much for the help and explanation.

dynasource replied:
thank you for your perseverance

dynasource reviewed (changes requested):
see comments about tests

nvanderburg committed:
Added removeValue method to ArrayHelper

samdark commented:
Merged. Thank you!

dynasource commented:
good work
